tests/mock_network/churn.rs

```diff
@@ -18,20 +18,256 @@ use routing::{
-/// iteration as `add_nodes_and_poll`.
+    let mut nodes = create_connected_nodes(&network, safe_section_size);
+    poll_and_resend(&mut nodes);
+    // Pause a node to act as non-responsive.
+    let mut rng = network.new_rng();
+    let non_responsive_index = gen_elder_index(&mut rng, &nodes);
+    let non_responsive_name = nodes[non_responsive_index].name();
+    info!(
+        "{:?} chosen as non-responsive.",
+        nodes[non_responsive_index].name()
+    );
+    let mut _non_responsive_node = None;
+
+    // Sending some user events to create a sequence of observations.
+    let mut responded = 0;
+    for i in 0..UNRESPONSIVE_WINDOW {
+        let event: Vec<_> = rng.gen_iter().take(100).collect();
+        nodes.iter_mut().for_each(|node| {
+            if node.name() == non_responsive_name {
+                // `chain_accumulator` gets reset during parsec pruning, which will reset the
+                // tracking of unresponsiveness as well. So this test has to assume there is no
+                // parsec pruning being carried out.
+                if responded < UNRESPONSIVE_WINDOW - UNRESPONSIVE_THRESHOLD - 1
+                    && rng.gen_weighted_bool(3)
+                {
+                    responded += 1;
+                } else {
+                    return;
+                }
+            }
+            let _ = node
+                .inner
+                .elder_state_mut()
+                .map(|state| state.vote_for_user_event(event.clone()));
+        });
+
+        // Required to avoid the case that the non-responsive node doesn't realize its removal,
+        // which blocks the polling infinitely.
+        if i == UNRESPONSIVE_THRESHOLD - 1 {
+            _non_responsive_node = Some(nodes.remove(non_responsive_index));
+        }
+
+        poll_and_resend(&mut nodes);
+    }
+
+    // Verify the other nodes saw the paused node and removed it.
+    for node in nodes.iter_mut().filter(|n| n.inner.is_elder()) {
+        expect_any_event!(node, Event::NodeLost(_));
+    }
+}
+
+// Parameters for the churn tests.
+//
+// The test run in three phases:
+// 1. In the grow phase nodes are only added.
+// 2. In the churn phase nodes are added and removed
+// 3. In the shrink phase nodes are only dropped
+struct Params {
+    // Network params
+    network: NetworkParams,
+    // The network starts with sections whose prefixes have these lengths. If empty, the network
+    // starts with just the root section with `elder_size` nodes.
+    initial_prefix_lens: Vec<usize>,
+    // When are messages sent during each iteration.
+    message_schedule: MessageSchedule,
+
+    // The grow phase lasts until the number of sections reaches this number.
+    grow_target_section_num: usize,
+    // Maximum number of iterations for the churn phase.
+    churn_max_iterations: usize,
+    // Probability that churn occurs for each iteration of the churn phase.
```

Comment:
better give an explaination how this factor being used to affect the churn rate?

Reply:
Isn't the existing comment enough? What would you suggest to add?

Reply:
specify the value shall be 0.0 - 1.0, and an example of 0.8 means 80% ?